### PR TITLE
Add distillation task to fine-tune local models 

### LINKED ISSUE #136: feat: Add distillation/training for text classification
I want to distil knowledge from a model, and it would be nice to have a single route from quickly experimenting with a model to distilling down to a smaller model.

For me, there are three things needed for this.

- data
    - provided by user
    - generated using a similar approach as https://github.com/argilla-io/synthetic-data-generator
- labels
    - sieves
- training
    - model2vec https://github.com/MinishLab/model2vec
    - setfit https://github.com/huggingface/setfit
    - fastfit https://github.com/IBM/fastfit

Potentially, we could add some post-processing of the labels by filtering based on entropy, label distribution or judge alignment.

### PR DESCRIPTION
**Pull request by rmitsch**
## Description

Adds distillation task to fine-tune local models. This is available as a new `Distillation` task. 
This PR implements distillation only for the `Classification` task - further task distillation implementations are out of scope for this PR.

  ```python
from sieves import Pipeline
from sieves.tasks import Distillation, DistillationFramework
from sieves.tasks.predictive import classification
 
engine = ...
docs =  ...
labels = ...
base_model_id = ...
  
pipe = Pipeline(
    [
        Classification(
            task_id="classifier",
            labels=labels,
            engine=engine,
        ),
        Distillation(
            target_task_id="classifier",
            base_model_id=base_model_id,
            framework=DistillationFramework.setfit,
            output_path=...,
        ),
    ]
)
          
  docs = list(pipe(docs))
  ```

## Related Issues

Resolves #136.

## Changes Made

- Add new `Distillation` task
- Implements distillation with `setfit` and `model2vec` for the classification task

## Checklist
- [x] Tests have been extended to cover changes in functionality
- [x] Existing and new tests succeed
- [x] Documentation updated (if applicable)
- [x] Related issues linked

### CODE REVIEW

File: sieves/tasks/postprocessing/distillation/core.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self._base_model_id = base_model_id
+        self._framework = framework
+        self._output_path = Path(output_path)
+        self._split_frac = split_fracs
+        self._init_kwargs = init_kwargs or {}
+        self._train_kwargs = train_kwargs or {}
+        self._seed = seed
+        self._threshold = threshold
+
+        self._target_task: PredictiveTask | None = None  # type: ignore[type-arg]
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/postprocessing/distillation/core.py:58]: It's `None` at initialization time. For distillation a `Task` has to have a reference to the target task, i.e. the task whose results they should distill. I don't want the user having to specify the task directly, as that would mean they'd have to do this:
```python
task1 = Classification(...)
pipe = Pipeline(
    [
        task1,
        DistillationTask(target_task=task1),
    ]
)
```
I.e. users couldn't init all tasks while creating the pipeline, so IMO usability would suffer a bit from that.

**Alternative 1** to this is to have a `Task` instance have a reference to the corresponding `Pipeline` instance. This is not the case with the current architecture, as there is no need for it otherwise. I could add that, but that seemed more blunt to me than alternative 2.

**Alternative 2:** set the reference to the target task to `None` at init time, and have the pipeline set the target task references post intialization. This means that the reference is `None` at first, but is never `None` when processing documents, since at that time the pipeline will have set the target reference.

What pattern would you prefer here?

File: sieves/tasks/postprocessing/distillation/core.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    @property
+    def target_task_id(self) -> str:
+        """Get target task ID.
+        :return: Target task ID.
+        """
+        return self._target_task_id
+
+    @property
+    def target_task(self) -> PredictiveTask | None:  # type: ignore[type-arg]
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/postprocessing/distillation/core.py:68]: See https://github.com/MantisAI/sieves/pull/138#discussion_r2062624096.

File: sieves/tasks/postprocessing/distillation/distillation_import.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+try:
+    import model2vec
+    import model2vec.train
+except ModuleNotFoundError:
+    model2vec = None
+
+    warnings.warn(_MISSING_WARNING.format(missing_dependency="model2vec"))
+
+
+__all__ = ["model2vec", "sentence_transformers", "setfit"]
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/postprocessing/distillation/distillation_import.py:41]: I've run into trouble when setting up `fastfit` training. Checked the repo and saw that the last update was six months ago with failing tests - it doesn't seem like it's actively maintained, so I'm hesitant to have it as dependency. 

Do you think it adds anything that `setfit`/`sentence-transformers`/`model2vec` can't do?

File: sieves/tasks/postprocessing/distillation/types.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from __future__ import annotations
+
+import enum
+
+
+class DistillationFramework(enum.Enum):
+    model2vec = "model2vec"
+    sentence_transformers = "sentence_transformers"
+    setfit = "setfit"
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/postprocessing/distillation/types.py:9]: See https://github.com/MantisAI/sieves/pull/138#discussion_r2060527241.

File: sieves/tasks/predictive/classification/core.py [hunk 5]
```diff
@@ -219,10 +223,10 @@ def _state(self) -> dict[str, Any]:
             "label_descriptions": self._label_descriptions,
         }
 
-    def to_dataset(self, docs: Iterable[Doc]) -> datasets.Dataset:
+    def to_hf_dataset(self, docs: Iterable[Doc], threshold: float = 0.5) -> datasets.Dataset:
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/predictive/classification/core.py:226]: Multi-label data is converted into a binary one-hot vector.

File: pyproject.toml [hunk 6]
```diff
@@ -45,7 +47,11 @@ all = [
     "ollama>=0.4,<1",
     "sentencepiece<1",
     "transformers>=4,<5",
-    "vllm>=0.8,<1"
+    "vllm>=0.8,<1",
+    # Distillation
+    "fast-fit>=1,<2",
+    "model2vec[train]>0.4,<0.5",
+    "sentence_transformers>=4,<5"
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply pyproject.toml]: Added `setfit` in https://github.com/MantisAI/sieves/pull/138/commits/d0ca6c4ddc881b4ec92d0ed0c157509725d4063a.

File: sieves/tasks/predictive/classification/core.py [hunk 7]
```diff
@@ -243,7 +248,96 @@ def generate_data() -> Iterable[dict[str, Any]]:
...
+
+        dataset_splits = self._split_dataset(hf_dataset, split_fracs, seed)
+        dataset_splits.save_to_disk(output_path / "data")
+
+        # Framework-specific distillation/fine-tuning logic using match-case
+        match distillation_framework:
+            case DistillationFramework.setfit:
+                model = setfit.SetFitModel.from_pretrained(
+                    base_model_id,
+                    multi_target_strategy="multi-output",
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/predictive/classification/core.py]: You're right, shouldn't be fixed. Changed in https://github.com/MantisAI/sieves/pull/138/commits/d0ca6c4ddc881b4ec92d0ed0c157509725d4063a to only be set in the multilabel case.

File: sieves/tasks/predictive/classification/core.py [hunk 8]
```diff
@@ -243,7 +248,96 @@ def generate_data() -> Iterable[dict[str, Any]]:
...
+                    save_strategy="epoch",
+                    load_best_model_at_end=True,
+                    **train_kwargs,
+                )
+
+                trainer = setfit.Trainer(
+                    model=model,
+                    args=args,
+                    train_dataset=hf_dataset,
+                    eval_dataset=hf_dataset,
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/predictive/classification/core.py]: Good catch, this was a typo.
1. Only doing a 2-way split now.
2. Using the splits now instead of the full dataset.

File: sieves/tasks/predictive/classification/core.py [hunk 9]
```diff
@@ -243,7 +248,96 @@ def generate_data() -> Iterable[dict[str, Any]]:
...
+                    metric="f1",
+                    column_mapping={"text": "text", "labels": "label"},
+                    metric_kwargs={"average": "macro"},
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/predictive/classification/core.py]: Same as above - valid point, only setting in the multilabel scenario now. See https://github.com/MantisAI/sieves/pull/138/commits/d0ca6c4ddc881b4ec92d0ed0c157509725d4063a.

File: sieves/tasks/predictive/classification/core.py [hunk 10]
```diff
@@ -243,7 +248,96 @@ def generate_data() -> Iterable[dict[str, Any]]:
...
+
+    def distill(
+        self,
+        base_model_id: str,
+        distillation_framework: DistillationFramework,
+        hf_dataset: datasets.Dataset,
+        init_kwargs: dict[str, Any],
+        train_kwargs: dict[str, Any],
+        output_path: Path | str,
+        split_fracs: tuple[float, float, float] = (0.8, 0.1, 0.1),
```
[the remark this answers is not in the document]
Comment (rmitsch) [reply sieves/tasks/predictive/classification/core.py]: Changed to .8, .2 in https://github.com/MantisAI/sieves/pull/138/commits/d0ca6c4ddc881b4ec92d0ed0c157509725d4063a.

**Comment by rmitsch** [comment]
@davidberenstein1957 All raised points should be addressed. Happy for you to give this another look.

**Comment by rmitsch** [comment]
@davidberenstein1957 Merging this in the next days if there are no further remarks from your side.
